design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,198 @@
+     because OAM platform builders or infrastructure teams can install any existing CRDs they
+     maintained to the platform as a `workload` or `trait`. Therefore, our solution needs to
+     accommodate pre-defined CRDs that OAM users bring.
+
+## Proposal
+The overall idea is for the applicationConfiguration controller to annotate critical information
+in the workload and trait CR it emits. In addition, we will provide a helper library so that
+trait controller developers can locate the resources they need with a simple function call. 
+Here is the list of changes that we propose.
+1. Add a `childResources` field in the  WorkloadDefinition. 
+Currently, a workloadDefinition is nothing but a shim of a real workload CRD. We propose to add
+an **optional** field called `childResources` to the schema of the workloadDefinition. We encourage
+workload owners to fill in this field when they register their controllers to the OAM system. 
+This is the way for them to declare the types of the Kubernetes resources their workload
+controller actually generates. In our example, the workload definition can claim to generate
+deployment and service child resources. 
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: WorkloadDefinition
+    metadata:
+       name: mydb.standard.oam.dev
+    spec:
+      definitionRef:
+         name: mydb.standard.oam.dev
+      childResources:
+         - groupVersion: apps/v1
+           kind: deployment
+         - groupVersion: core/v1
+           kind: service      
+    ```
+
+2. ApplicationConfig controller add annotations containing workload child resource information to
+the emitted `workload` CR. For each `workload` CR, the applicationConfig controller can fetch the
+corresponding `workloadDefinition` CR following an 
+[OAM convention](https://github.com/oam-dev/spec/blob/master/3.workload.md#definitionref). The
+conventions says that the name of the `workloadDefinition` CR is the name of the `workload` CRD
+it refers to. For example, the name of the `workloadDefinition` CR that refers to a
+`containerizedworkloads.core.oam.dev` CRD is exactly `containerizedworkloads.core.oam.dev` as
+well. In our example, the applicationConfig controller will emit the following `workload` CR with all
+ its child resources GVK in its metadata.
+    ```yaml
+    apiVersion: standard.oam.dev/v1alpha2
+    kind: Mydb
+    metadata:
+      name: mydb-example
+      annotations:
+        core.oam.dev/childResources: "{ apps/v1.deployment, core/v1.service}"
+    ```
+
+3. ApplicationConfig controller add annotations containing the workload information to the emitted
+`trait` CR. The application controller actually generates the workload CR so it has all the
+information. In our example, the applicationConfig controller will emit the following `trait` CR with
+the GVK and name of the workload the `ManualScalerTrait` is applied to.
```

Comment:
I wonder about the best way to handle situations in which a trait is defined for a workload in the `ApplicationConfiguration`, but the trait is not actually able to modify the rendered child resources. For instance, the manual scaler trait can only modify workload types that render out a `Deployment` right now. The model described here would render out the workload components then the `ManualScalerTrait` would try to modify the `Deployment`. If the workload did not produce a `Deployment`, the `ManualScalerTrait` controller would likely just not do anything here. This might be the right strategy (i.e. if you use a trait that does not apply to the workload it is basically a no-op). However, we could also do a similar thing that you propose for `WorkloadDefinition` in `TraitDefinition` by adding a `appliesTo` field (not sure that is the best name) with a slice of object types it knows how to modify. So for the `ManualScalerTrait` it could look like this:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: manualscalertraits.core.oam.dev
spec:
  definitionRef:
    name: manualscalertraits.core.oam.dev
  appliesTo:
    - apiVersion: apps/v1
      kind: Deployment
    - apiVersion: apps/v1
      kind: StatefulSet # not currently supported but useful for illustration purposes
```

Following this pattern would allow the `ApplicationConfiguration` controller to decide if a configuration is valid, and it could refuse to render out workloads / traits if it is not. I am not sure if this is more desirable or not, but it could guard against situations where a user thinks they are deploying one configuration, but another one is actually being created (i.e. we won't render a workload if the trait is not compatible).

Reply:
I like this idea and it plays well with the overall trait <-> workload compatibility scheme although the `traitDefinition` CRD already has an `appliesToWorkloads ` field.
To be more specific, I think this is part of the other problem we ought to solve. Namely, how can we know if a `trait` can be applied to a `workload`.  I think we should definitely incorporate this idea when we get to implement https://github.com/oam-dev/spec/issues/153.  The overall idea is that we will add some annotation keys to a component to indicate its runtime characteristics. We can then use the `appliesToWorkloads ` field in the `traitDefinition` to see if the trait can apply to the component instance. I think we can combine your idea with that scheme.  We will probably use a validatingMutator to implement it.